File: radar-commons/src/main/java/org/radarbase/data/RemoteSchemaEncoder.java
package org.radarbase.data;

import java.io.IOException;
import org.apache.avro.Schema;
import org.apache.avro.SchemaValidationException;
import org.apache.avro.generic.GenericData;
import org.apache.avro.specific.SpecificData;
import org.apache.avro.specific.SpecificRecord;
import org.radarbase.producer.rest.AvroDataMapper;
import org.radarbase.producer.rest.AvroDataMapperFactory;
import org.radarbase.producer.rest.ParsedSchemaMetadata;

/**
 * Encodes data according to an Avro schema to the format and schema of the server.
 */
public class RemoteSchemaEncoder implements AvroEncoder {
    private final boolean binary;

    /**
     * Schema encoder.
     * @param binary true if the server wants binary encoding, false if it wants JSON encoding.
     */
    public RemoteSchemaEncoder(boolean binary) {
        this.binary = binary;
    }

    @Override
    public <T> AvroWriter<T> writer(Schema schema, Class<? extends T> clazz) {
        return new SchemaEncoderWriter<>(schema, clazz);
    }

    private class SchemaEncoderWriter<T> implements AvroWriter<T> {
        private final AvroEncoder recordEncoder;
        private AvroEncoder.AvroWriter<Object> encoder;
        private final boolean isGeneric;
        private ParsedSchemaMetadata serverSchema;
        private AvroDataMapper mapper;
        private final Schema schema;

        SchemaEncoderWriter(Schema schema, Class<? extends T> clazz) {
            this.schema = schema;

            GenericData genericData;
            if (SpecificRecord.class.isAssignableFrom(clazz)) {
                genericData = new SpecificData(RemoteSchemaEncoder.class.getClassLoader());
                isGeneric = false;
            } else {
                genericData = new GenericData(RemoteSchemaEncoder.class.getClassLoader());
                isGeneric = true;
            }
            recordEncoder = new AvroDatumEncoder(genericData, binary);
        }

        @Override
        public byte[] encode(T object) throws IOException {
            return encoder.encode(mapper.convertAvro(object));
        }

        @Override
        public final void setReaderSchema(ParsedSchemaMetadata readerSchema)
                throws SchemaValidationException {
            if (this.serverSchema != null
                    && readerSchema.getSchema().equals(this.serverSchema.getSchema())) {
                return;
            }
            try {
                if (!isGeneric) {
                    this.mapper = AvroDataMapperFactory.IDENTITY_MAPPER;
                    encoder = recordEncoder.writer(schema, Object.class);
                } else {
                    this.mapper = AvroDataMapperFactory.get()
                            .createMapper(schema, readerSchema.getSchema(),
                                    null);
                    encoder = recordEncoder.writer(readerSchema.getSchema(), Object.class);
                }
                this.serverSchema = readerSchema;
            } catch (IOException ex) {
                throw new IllegalStateException("Cannot construct Avro writer", ex);
            }
        }

        @Override
        public ParsedSchemaMetadata getReaderSchema() {
            return serverSchema;
        }
    }
}

File: radar-commons/src/main/java/org/radarbase/data/RemoteSchemaEncoder.kt
package org.radarbase.data

import org.apache.avro.Schema
import org.apache.avro.generic.GenericData
import org.apache.avro.specific.SpecificData
import org.apache.avro.specific.SpecificRecord
import org.radarbase.data.AvroEncoder.AvroWriter
import org.radarbase.producer.avro.AvroDataMapper
import org.radarbase.producer.avro.AvroDataMapperFactory
import org.radarbase.producer.avro.AvroDataMapperFactory.validationException
import java.io.IOException

/**
 * Encodes data according to an Avro schema to the format and schema of the server.
 *
 * @param binary true if the server wants binary encoding, false if it wants JSON encoding.
 */
class RemoteSchemaEncoder(
    private val binary: Boolean,
) : AvroEncoder {
    override fun <T: Any> writer(schema: Schema, clazz: Class<out T>, readerSchema: Schema): AvroWriter<T> {
        return SchemaEncoderWriter(binary, schema, clazz, readerSchema)
    }

    class SchemaEncoderWriter<T: Any>(
        binary: Boolean,
        schema: Schema,
        clazz: Class<out T>,
        readerSchema: Schema,
    ) : AvroWriter<T> {
        private val recordEncoder: AvroEncoder
        private val encoder: AvroWriter<Any>
        private val isGeneric: Boolean
        private val mapper: AvroDataMapper

        init {
            if (schema.type !== Schema.Type.RECORD) throw validationException(schema, readerSchema, "Can only map records.")
            val genericData: GenericData
            val classLoader = Thread.currentThread().contextClassLoader
            val useReaderSchema: Schema
            if (SpecificRecord::class.java.isAssignableFrom(clazz)) {
                genericData = SpecificData(classLoader)
                useReaderSchema = schema
                isGeneric = false
            } else {
                genericData = GenericData(classLoader)
                useReaderSchema = readerSchema
                isGeneric = true
            }
            recordEncoder = AvroDatumEncoder(genericData, binary)
            mapper = AvroDataMapperFactory.createMapper(schema, useReaderSchema, null)
            encoder = recordEncoder.writer(useReaderSchema, Any::class.java)
        }

        @Throws(IOException::class)
        override fun encode(`object`: T): ByteArray = encoder.encode(
            requireNotNull(mapper.convertAvro(`object`)) {
                "Cannot map $`object` to Avro"
            }
        )
    }
}
